pushy/src/main/java/com/relayrides/pushy/apns/ApnsClient.java

```diff
@@ -826,46 +802,34 @@ protected AuthenticationTokenSupplier getAuthenticationTokenSupplierForTopic(fin
         final ChannelPromise connectionReadyPromise = this.connectionReadyPromise;
 
         if (connectionReadyPromise != null && connectionReadyPromise.isSuccess() && connectionReadyPromise.channel().isActive()) {
-            if (!connectionReadyPromise.channel().isWritable()) {
+            final Channel channel = connectionReadyPromise.channel();
+
+            if (!channel.isWritable()) {
                 return new FailedFuture<>(GlobalEventExecutor.INSTANCE, CLIENT_BUSY_EXCEPTION);
             }
 
-            final Promise<PushNotificationResponse<ApnsPushNotification>> responsePromise;
-
-            if (promise != null) {
-                responsePromise = promise;
-            } else {
-                responsePromise = new DefaultPromise<>(connectionReadyPromise.channel().eventLoop());
-            }
+            final Promise<PushNotificationResponse<ApnsPushNotification>> responsePromise =
+                    new DefaultPromise<>(channel.eventLoop());
 
-            connectionReadyPromise.channel().eventLoop().submit(new Runnable() {
+            channel.eventLoop().submit(new Runnable() {
 
                 @Override
                 public void run() {
-                    if (ApnsClient.this.responsePromises.containsKey(notification)) {
-                        responsePromise.setFailure(new IllegalStateException(
-                                "The given notification has already been sent and not yet resolved."));
-                    } else {
-                        // We want to do this inside the channel's event loop so we can be sure that only one thread is
-                        // modifying responsePromises.
-                        ApnsClient.this.responsePromises.put(notification, responsePromise);
-                    }
-                }
-            });
+                    final ApnsClientHandler handler = channel.pipeline().get(ApnsClientHandler.class);
 
-            connectionReadyPromise.channel().writeAndFlush(notification).addListener(new GenericFutureListener<ChannelFuture>() {
-
-                @Override
-                public void operationComplete(final ChannelFuture future) throws Exception {
-                    if (future.isSuccess()) {
-                        ApnsClient.this.metricsListener.handleNotificationSent(ApnsClient.this, notificationId);
-                    } else {
-                        log.debug("Failed to write push notification: {}", notification, future.cause());
+                    try {
+                        handler.setResponsePromiseForNotification(notification, responsePromise);
```

Comment:
The best idea I've come up with to make this a little more atomic is to actually write some kind of `PushNotificationAndResponsePromise` object instead of just the push notification. I think the awkwardness there may outweigh the benefit, though.